### PR TITLE
Move cumprod and cumsum to Aten(CPU)

### LINKED ISSUE #24668: Migrate `_cumprod` from the TH to Aten (CPU)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### PR DESCRIPTION
**Pull request by XiaobingSuper**
This PR is about move cumprod and cumsum to Aten.
Test script:
```
import torch
import torch.nn as nn
import time

torch.manual_seed(0)

def _time():
    return time.time()

device = "cpu"

#torch.set_num_threads(1)

#warm up
for n in [10, 300]:
    input = torch.randn(n, n, n, requires_grad=False, device=device)
    input = input * 0.01 + 1
    for dim in range(input.dim()):
        for i in range(100):
            #output = input.cumsum(dim)
            output = input.cumprod(dim)

for n in [10, 300]:
    input = torch.randn(n, n, n, requires_grad=False, device=device)
    input = input * 0.01 + 1
    for dim in range(input.dim()):
        fwd_t = 0
        for i in range(1000):
            t1 = _time()
            #output = input.cumsum(dim)
            output = input.cumprod(dim)
            t2 = _time()
            fwd_t = fwd_t + (t2 -t1)
        fwd_avg = fwd_t / 1000 * 1000
        print("size = (%d, %d, %d); reduce dim=%d; compute time is %.4f(ms)" % (n, n, n, dim, fwd_avg))
```
Test device: **skx-8180**.
Performance:
```
cumsum:
Before:
size = (10, 10, 10); reduce dim=0; compute time is 0.0108(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0097(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0096(ms)
size = (300, 300, 300); reduce dim=0; compute time is 220.2608(ms)
size = (300, 300, 300); reduce dim=1; compute time is 252.3499(ms)
size = (300, 300, 300); reduce dim=2; compute time is 71.8618(ms)
number_threads = 1:
size = (10, 10, 10); reduce dim=0; compute time is 0.0125(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0113(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0110(ms)
size = (300, 300, 300); reduce dim=0; compute time is 212.4133(ms)
size = (300, 300, 300); reduce dim=1; compute time is 245.6995(ms)
size = (300, 300, 300); reduce dim=2; compute time is 67.4413(ms)
After:
size = (10, 10, 10); reduce dim=0; compute time is 0.0065(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0063(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0053(ms)
size = (300, 300, 300); reduce dim=0; compute time is 36.0139(ms)
size = (300, 300, 300); reduce dim=1; compute time is 36.0776(ms)
size = (300, 300, 300); reduce dim=2; compute time is 21.0111(ms)
number_threads = 1:
size = (10, 10, 10); reduce dim=0; compute time is 0.0053(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0052(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0051(ms)
size = (300, 300, 300); reduce dim=0; compute time is 81.8831(ms)
size = (300, 300, 300); reduce dim=1; compute time is 88.5687(ms)
size = (300, 300, 300); reduce dim=2; compute time is 54.9922(ms)
```
```
cumprod:
Before:
size = (10, 10, 10); reduce dim=0; compute time is 0.0096(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0088(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0088(ms)
size = (300, 300, 300); reduce dim=0; compute time is 221.2601(ms)
size = (300, 300, 300); reduce dim=1; compute time is 249.7894(ms)
size = (300, 300, 300); reduce dim=2; compute time is 71.5182(ms)
number_threads = 1:
size = (10, 10, 10); reduce dim=0; compute time is 0.0100(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0093(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0093(ms)
size = (300, 300, 300); reduce dim=0; compute time is 207.6287(ms)
size = (300, 300, 300); reduce dim=1; compute time is 241.6693(ms)
size = (300, 300, 300); reduce dim=2; compute time is 66.2977(ms)
After:
size = (10, 10, 10); reduce dim=0; compute time is 0.0063(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0062(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0053(ms)
size = (300, 300, 300); reduce dim=0; compute time is 36.4283(ms)
size = (300, 300, 300); reduce dim=1; compute time is 38.1139(ms)
size = (300, 300, 300); reduce dim=2; compute time is 20.9140(ms)
number_threads =1:
size = (10, 10, 10); reduce dim=0; compute time is 0.0052(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0052(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0050(ms)
size = (300, 300, 300); reduce dim=0; compute time is 82.6926(ms)
size = (300, 300, 300); reduce dim=1; compute time is 90.1265(ms)
size = (300, 300, 300); reduce dim=2; compute time is 55.0196(ms)
```
Fix https://github.com/pytorch/pytorch/issues/24668, https://github.com/pytorch/pytorch/issues/24669.

### CODE REVIEW

**Comment by ngimel** [comment]
cc @anjali411

**Comment by ngimel** [comment]
#33137 and this generally look good. 
- There's a lot of code repetition between them (the only difference is init value, and + or *), so consider templating. 
- Don't hardcode GRAIN_SIZE, better use internal::GRAIN_SIZE. Since in your case each thread is doing a fair amount of work, GRAIN_SIZE/N/n seems reasonable. 
- Please also provide single-threaded benchmarks (you can use `torch.set_num_threads(1)`)

**Comment by XiaobingSuper** [comment]
There has a performance issue for cumprod, working in progress.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 1]
```diff
@@ -180,6 +182,89 @@ static TensorIterator make_reduction(
   return TensorIterator::reduce_op(viewed_result1, viewed_result2, self.to(dtype));
 }
 
+template<typename scalar_t, typename func_t>
+void _cumop_out_cpu_template(Tensor& result, const Tensor& self,
+    int64_t dim, scalar_t init_v, func_t && op) {
+  auto wrap_dim = maybe_wrap_dim(dim, self.dim());
+  TORCH_CHECK((0 <= wrap_dim && wrap_dim < self.dim())
+              || (0 == wrap_dim && wrap_dim == self.dim()) , "invaild dim to reduce");
```
Comment (anjali411) [inline aten/src/ATen/native/ReduceOps.cpp]: I don't think this check is required. maybe_wrap_dim already does that

File: aten/src/ATen/native/ReduceOps.cpp [hunk 2]
```diff
@@ -180,6 +182,89 @@ static TensorIterator make_reduction(
...
+    result.resize_as_(self);
+  }
+  if (self.numel() == 0) {
+    return;
+  }
+  const auto input_shape = self.sizes();
+  const auto input_ndim = self.dim();
+
+  if (input_ndim == 0) {
+    result.copy_(self);
```
Comment (anjali411) [inline aten/src/ATen/native/ReduceOps.cpp]: '''result.fill_(self)'''

**Comment by anjali411** [comment]
I think we should have a general function with equivalent logic to TENSOR_DIM_APPLY2. you can check my PR here where I porter TENSOR_DIM_APPLY3 for cummax, cummin https://github.com/pytorch/pytorch/pull/32537/files it would be nice if you can add your function to the new file I am adding through this PR

**Review by anjali411 — CHANGES_REQUESTED** [review]
I think we should have a general function with equivalent logic to TENSOR_DIM_APPLY2. you can check my PR here where I porter TENSOR_DIM_APPLY3 for cummax, cummin https://github.com/pytorch/pytorch/pull/32537/files it would be nice if you can add your function to the new file I am adding through this PR

**Comment by XiaobingSuper** [comment]
@anjali411, @ngimel , @VitalyFedyunin , DIM_APPlY is not be parallelized, it has a bad performance compare with this PR's method even for non_contiguous tensor(amost),  please see the performance of cumsum when input is discontiguous :
**Before:**
```
size = (10, 10, 10); reduce dim=0; compute time is 0.0096(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0088(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0087(ms)
size = (300, 300, 300); reduce dim=0; compute time is 220.7853(ms)
size = (300, 300, 300); reduce dim=1; compute time is 249.2653(ms)
size = (300, 300, 300); reduce dim=2; compute time is 66.2485(ms)
number_threads =1:
size = (10, 10, 10); reduce dim=0; compute time is 0.0095(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0089(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0089(ms)
size = (300, 300, 300); reduce dim=0; compute time is 224.9825(ms)
size = (300, 300, 300); reduce dim=1; compute time is 251.4869(ms)
size = (300, 300, 300); reduce dim=2; compute time is 66.5189(ms)
```
After:
```
size = (10, 10, 10); reduce dim=0; compute time is 0.0062(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0061(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0061(ms)
size = (300, 300, 300); reduce dim=0; compute time is 90.9508(ms)
size = (300, 300, 300); reduce dim=1; compute time is 86.9578(ms)
size = (300, 300, 300); reduce dim=2; compute time is 27.2669(ms)

number_threads =1:
size = (10, 10, 10); reduce dim=0; compute time is 0.0056(ms)
size = (10, 10, 10); reduce dim=1; compute time is 0.0055(ms)
size = (10, 10, 10); reduce dim=2; compute time is 0.0056(ms)
size = (300, 300, 300); reduce dim=0; compute time is 111.3135(ms)
size = (300, 300, 300); reduce dim=1; compute time is 106.2396(ms)
size = (300, 300, 300); reduce dim=2; compute time is 77.9179(ms)
```

**Comment by ngimel** [comment]
Agreed, you don't need to use dim_apply. Did you fix issues that you have for cumprod performance? If so, then there are small things left to fix
1) add scalar handling before maybe_wrap_dim because maybe_wrap_dim will error out on scalar, and you want to be able to do the following, it is currently supported 
```
a=torch.tensor(1) # create scalar tensor with 0 dims
a.cumprod(dim=0)
```
2) remove TORCH_CHECK on the `wrap_dim` as @anjali411 suggests, it is done by maybe_wrap_dim
3) add a test for cumsum/cumprod on a scalar, apparently it is not tested.

**Comment by XiaobingSuper** [comment]
Using TensorItereator can get a more performance improvement than parallel_for(will have **~2x** improvement) when dim != input.dim() - 1. TODO: solve the performance degradation for dim == input.dim() - 1.

**Comment by ngimel** [comment]
@XiaobingSuper what approach are you using with TensorIterator? Is it similar to Scatter/Gather (restriding original tensor so that the stride is 0 for reduced dimension, capturing the op over that dimension in the function that is passed to iterator)? https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cpu/ScatterGatherKernel.cpp. 
If TensorIterator still does not give you performance, you can continue with your original approach, it still offers benefits compared to current implementation. 
Let me know if you need help or when your PR is ready for review.

**Comment by XiaobingSuper** [comment]
@ngimel, I revert TensorIterator patch, just using original approch now, cumprod still have a performance degradation for single thread case.

**Comment by ngimel** [comment]
Hm, it's so weird that cumsum and cumprod perf is different, the code is literally the same.

File: aten/src/ATen/native/ReduceOps.cpp [hunk 3]
```diff
@@ -183,6 +185,86 @@ static TensorIterator make_reduction(
...
+      );
+
+  auto self_ptr = self.data_ptr<scalar_t>();
+  auto out_ptr = temp_tensor.data_ptr<scalar_t>();
+  auto offset = N * n;
+  parallel_for(0, M, internal::GRAIN_SIZE, [&](int64_t begin, int64_t end) {
+    auto s_index =  offset * begin;
+    for (int64_t f = begin; f < end; ++f) {
+      for (auto j = 0; j < N; j++ ) {
+        scalar_t cum_number = init_v;
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/ReduceOps.cpp]: You might get precision problem here, as original code had accumulator of `accreal` type

File: aten/src/ATen/native/ReduceOps.cpp [hunk 4]
```diff
@@ -183,6 +185,86 @@ static TensorIterator make_reduction(
   return TensorIterator::reduce_op(viewed_result1, viewed_result2, self.to(dtype));
 }
 
+template<typename scalar_t, typename func_t>
+static void _cumop_out_cpu_template(Tensor& result, const Tensor& self,
```
Comment (VitalyFedyunin) [inline aten/src/ATen/native/ReduceOps.cpp]: Wild guess, try force inlining, maybe it will solve perf problem.

File: aten/src/ATen/native/ReduceOps.cpp (hunk 3 above)
Comment (VitalyFedyunin) [reply aten/src/ATen/native/ReduceOps.cpp]: Discussed with @ngimel, we are fine using scalar_t here for acc.

**Comment by VitalyFedyunin** [comment]
I want to support @ngimel 's proposal to have quick path in case of `dim == input.dim()-1` this should remove performance degradation.

**Comment by ngimel** [comment]
The reason for the slowdown is denormals appearing at the end of the long chain of multiplying numbers less than 1. If the inputs to the benchmark are prepared as 
```
input = torch.rand(n, n, n, requires_grad=False, device=device, dtype=torch.float)
    input = input * 0.01 + 1
```
there's no difference in performance between cumsum and cumprod.

**Comment by XiaobingSuper** [comment]
> The reason for the slowdown is denormals appearing at the end of the long chain of multiplying numbers less than 1. If the inputs to the benchmark are prepared as
> 
> ```
> input = torch.rand(n, n, n, requires_grad=False, device=device, dtype=torch.float)
>     input = input * 0.01 + 1
> ```
> 
> there's no difference in performance between cumsum and cumprod.

@ngimel , yes, it works. Thanks!

**Comment by ngimel** [comment]
@XiaobingSuper can you please resolve a conflict?

**Comment by XiaobingSuper** [comment]
@ngimel, @VitalyFedyunin , C@anjali411, there have some updates:
1. Using TensorIterator as implementation method.
2. Updata performance data.
3. Using acctype to make computation.

**Review by VitalyFedyunin — APPROVED** [review]

**Comment by zou3519** [comment]
Doesn't look BC-breaking because the code does use an acctype for accumulation
